pupil_src/shared_modules/calibration_choreography/single_marker_plugin.py

```diff
@@ -0,0 +1,383 @@
+
+        elif isinstance(state, MarkerWindowStateIdle):
+            self.__marker_window.show_marker(
+                self._FIXED_MARKER_POSITION, should_animate=should_animate
+            )
+
+        elif isinstance(state, MarkerWindowStateAnimatingInMarker):
+            pass  # No-op
+
+        elif isinstance(state, MarkerWindowStateShowingMarker):
+            assert self.marker_mode != SingleMarkerMode.MANUAL
+            should_save_ref_marker = True
+
+        elif isinstance(state, MarkerWindowStateAnimatingOutMarker):
+            pass  # No-op
+
+        else:
+            raise UnhandledMarkerWindowStateError(state)
+
+        if should_save_ref_marker and ref_marker is not None and stop_marker is None:
+            ref = {}
+            ref["norm_pos"] = ref_marker["norm_pos"]
+            ref["screen_pos"] = ref_marker["img_pos"]
+            ref["timestamp"] = frame.timestamp
+            self.ref_list.append(ref)
+
+        # Update UI
+        self.__marker_window.draw_window()
+        self.status_text = self._FIXED_MARKER_POSITION if self.is_active else None
+
+    def gl_display(self):
+        """
+        use gl calls to render
+        at least:
+            the published position of the reference
+        better:
+            show the detected postion even if not published
+        """
+
+        if not self.is_active:
+            return
+
+        for marker in self.__previously_detected_markers:
+            # draw the largest ellipse of all detected markers
+            e = marker["ellipses"][-1]
+            pts = cv2.ellipse2Poly(
+                (int(e[0][0]), int(e[0][1])),
+                (int(e[1][0] / 2), int(e[1][1] / 2)),
+                int(e[-1]),
+                0,
+                360,
+                15,
+            )
+            draw_polyline(pts, color=RGBA(0.0, 1.0, 0, 1.0))
+            if len(self.__previously_detected_markers) > 1:
+                draw_polyline(pts, 1, RGBA(1.0, 0.0, 0.0, 0.5), line_type=gl.GL_POLYGON)
+
+            # draw indicator on the stop marker(s)
+            if marker["marker_type"] == "Stop":
+                e = marker["ellipses"][-1]
+                pts = cv2.ellipse2Poly(
+                    (int(e[0][0]), int(e[0][1])),
+                    (int(e[1][0] / 2), int(e[1][1] / 2)),
+                    int(e[-1]),
+                    0,
+                    360,
+                    360 // self._STOP_MARKER_FRAMES_NEEDED_TO_STOP,
+                )
+                indicator = (
+                    [e[0]]
+                    + pts[self.__auto_stop_tracker.detected_count :].tolist()
+                    + [e[0]]
+                )
+                draw_polyline(
+                    indicator, color=RGBA(8.0, 0.1, 0.1, 0.8), line_type=gl.GL_POLYGON
+                )
+
+    ### Internal
+
+    def _perform_start(self):
+        # TODO: Is this check duplicated for all calibration choreographies?
+        if not self.g_pool.capture.online:
```

Comment:
@papr Could you please clarify why this check exists and if it makes sense for all choreography plugins, or only `screen marker` and `single marker`?

Reply:
Necessary for all choreos that rely on having a world video, i.e. screen marker, single marker, and natural features.